[Dilys Pang] iP

Dilysss opened the pull request:
# DUKE BOT 🤖 

> "The way to get started is to quit talking and begin doing." - Walt Disney

🎉 The best todo list you can find! 🎉 

## Why should you use it?
- *Simple* to learn
- *Aesthetically* pleasing
- *Functional* features

## Just follow these simple steps to start using Duke! ✨ 
1. Click [here](https://github.com/Dilysss/ip/releases/tag/A-Jar) to download duke.jar
2. Unzip and launch the app
3. Begin adding your tasks!

### UNLOCK ALL THESE AMAZING FEATURES
- [x] Add Todos, Events and Deadlines 📋 
- [ ]  Mark and unmark tasks ✅
- [ ]  Delete tasks ❌ 
- [ ]  Search for tasks 🔍

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Dilysss committed:
Level-1

Dilysss committed:
Level-2

Dilysss committed:
Level-3

Dilysss committed:
Level-4

Dilysss committed:
Level-5

Dilysss committed:
Level-6

Dilysss committed:
Level-6 edited

Dilysss committed:
Added unmark and horizontal lines, edited logic

Dilysss committed:
Amended automated text UI testing

Dilysss committed:
Add Level-7 and Level-8

Dilysss committed:
Modify Level-8

Dilysss committed:
Add Packages

Dilysss committed:
Made changes to Save

Dilysss committed:
Merge branch 'add-gradle-support'

Dilysss committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Dilysss committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

Dilysss committed:
Added JUnit Tests

Dilysss committed:
Merge remote-tracking branch 'origin/master'

Dilysss committed:
Updated level 9

Dilysss committed:
Updated level 9

Dilysss committed:
Updated storage, task and todos

parnikkapore commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,47 @@
+ */
+public class Deadlines extends Task {
+
+    protected String description;
```
Should this line be removed? This field does not even seem to be used within this file, and it looks like the description is supposed to be stored in Task.

parnikkapore commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public static void toDo() throws DukeException {
+        Parser parser = new Parser(ui, tasklist, storage);
```
I like how you are using dependency injection early on in your career!

parnikkapore commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,47 @@
+     * @return Formatted string.
+     */
+    private String dateStr() {
+        String formatDateTime = this.dateTime.format(DateTimeFormatter.ofPattern("dd MMM yyyy HH:mm"));
```
Should this be named `formattedDateTime`? I think the current name sounds like the name of a method.

parnikkapore commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,47 @@
+     *
+     * @return Formatted string.
+     */
+    private String dateStr() {
```
Should this be named `getDateString`, as is typical for getters? I'm pretty sure that this is also implied by the [coding standard](https://se-education.org/guides/conventions/java/index.html#naming) saying that method names need to be verbs.

This also applies to other getters specified in `Task`.

parnikkapore commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    public static void toDo() throws DukeException {
```
I'm not sure why toDo() needs to be a separate function. Is it to support multiple bots embedded within Duke (yours truly have done something similar)?

parnikkapore commented on src/main/java/duke/Events.java:
```diff
@@ -0,0 +1,46 @@
+ * Represents an events task to be added to Duke with a event date.
+ */
+public class Events extends Task {
+    protected String eventTime;
+    private LocalDateTime dateTime;
```
Which of these two variables are storing the time of the event?

parnikkapore commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,189 @@
+            storage.save(tasklist);
+        } else if (userInput.equals("todo")) {
+            throw new DukeException("OOPS!!! The description of a todo cannot be empty.");
+        } else if (userInput.equals("deadline")) {
```
I'm not sure I like the fact that the branch checking for a deadline command without details is separated from the branch handling a deadline command with details.

parnikkapore commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,189 @@
+            throw new DukeException("OOPS!!! The description of a deadline cannot be empty.");
+        } else if (userInput.equals("event")) {
+            throw new DukeException("OOPS!!! The description of a event cannot be empty.");
+        } else if (userInput.length() > 8 && (userInput.substring(0, 8)).equals("deadline")) {
```
This branch will match the input `deadlinesNextWeek`. Perhaps use the technique you discovered on line 54?

```suggestion
        } else if (userInput.startsWith("deadline ")) {
```

parnikkapore commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,189 @@
+     * @param str The string containing task to be added.
+     */
+    public static void addDeadline(String str) {
+        int k = 9;
```
Perhaps move this to above line 121 so that it is clear this is a loop counter?

It's also unclear to me why the counter starts at 9.

A similar thing applies to `addEvent`.

parnikkapore commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,189 @@
+        int taskToMark = 0;
+        String strTaskToMark = "";
+
+        for (int j = 5; j < task.length(); j++) {
+            strTaskToMark = strTaskToMark + task.charAt(j);
+        }
```
Is there any reason you did it this way instead of using `task.substring()`?

parnikkapore commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,87 @@
+        ArrayList<Task> taskArr = new ArrayList<>();
+        File file = new File(filePath);
+        try {
+
```
Why is there a blank line here?

parnikkapore reviewed (changes requested):
Great job adding several new features to the bot. Aside from the naming issues, I'm also a bit concerned about the input parsing. In context though, that isn't too big a deal. Congrats on surviving the mod thus far!

Dilysss committed:
Updated storage, task and todos

Dilysss committed:
Finish Level-10

elizabethhky commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,47 @@
+     * @return Data to be saved.
+     */
+    @Override
+    public String savedData() {
```
Perhaps the method can start with a verb for clearer understanding of what the method does? 
```suggestion
    public String printSavedData() {
```

elizabethhky commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Returns string for deadline to be displayed on Duke.
+     *
+     * @return string of deadline.
```
I noticed that there are some method descriptions that have a capitalised returned object. 
e.g For getDateStr(), the method description says "@return Formatted string." But here, it is "@return string of deadline." 

Maybe you would like to standardise that all return objects be capitalised?

elizabethhky commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    public String getResponse(String input) throws DukeException {
```
Should there be a method description for getResponse?

elizabethhky commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,8 @@
+package duke;
+
+public class DukeException extends Exception {
```
Should there be a class description for DukeException?

elizabethhky commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Storage {
```
Should there be a class description for Storage?

elizabethhky commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+    private String filePath;
+
+    public Storage(String filePath) {
+
```
Would it be better to delete this empty line?

elizabethhky commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,89 @@
+            throw new DukeException("New file can't be created");
+        }
+
+        try {
+            File dukeFile = new File(filePath);
+            Scanner sc = new Scanner(dukeFile);
+
+            while (sc.hasNextLine()) {
+                String[] strArr = sc.nextLine().split("\\|");
+                if (strArr[0].contains("T")) {
+                    taskArr.add(new ToDos(strArr[2]));
+                }
+                if (strArr[0].contains("E")) {
+                    String str = strArr[3];
+                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern(" dd MMM yyyy HH:mm");
+                    LocalDateTime ldt = LocalDateTime.parse(str, formatter);
+                    taskArr.add(new Events(strArr[2], ldt));
+                }
+                if (strArr[0].contains("D")) {
+                    String str = strArr[3];
+                    DateTimeFormatter formatter = DateTimeFormatter.ofPattern(" dd MMM yyyy HH:mm");
+                    LocalDateTime ldt = LocalDateTime.parse(str, formatter);
+                    taskArr.add(new Deadlines(strArr[2], ldt));
+                }
+            }
+            sc.close();
+            return new TaskList(taskArr);
```
I really like this coding style! 😄

elizabethhky commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,37 @@
+package duke;
+
+/**
+ * Represents task object
```
I noticed that there are some places where the comments do not end with a full stop. (e.g in this Task.java file and TaskList.java) Would it be better to add a full stop at the end of every comment?

elizabethhky commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Prints list of tasks.
+     */
+    //public String printList(ArrayList<Task> tasks) {
```
I noticed that there are quite a lot of commented out code in some files. Would it be neater to delete them if you do not plan on using them again?

elizabethhky commented on src/test/java/duke/DeadlinesTest.java:
```diff
@@ -0,0 +1,17 @@
+public class DeadlinesTest {
+
+    @Test
+    public void deadlinesTest(){
```
For your test files, there is another way to name your test methods following this format - `featureUnderTest_testScenario_expectedBehavior()` 

Perhaps you can follow the format for clearer understanding of each testcase. Otherwise, I think it is fine to leave it as it is as well! 😊

elizabethhky commented on src/main/java/duke/Deadlines.java:
```diff
@@ -0,0 +1,47 @@
+package duke;
```
Would it be better to separate out your classes into more packages, e.g duke.tasks, duke.exception?

elizabethhky reviewed (commented):
Well written code! 👍  There are some minor tweaks to be made but overall good job on following the coding standard closely 😄

Dilysss committed:
Added missing javadocs, removed commented code

Dilysss committed:
Add Statistic Extension

No methods to calculate the number of todos, events and deadlines
respectively.

Provides a way to leverage statistics about the tasks managed by the
Duke Bot.

Create methods to calculate the number of todos, events and deadlines
by handling these tasks within a same method in the Parser class.

Handling these tasks within the same method helps to reduce code duplication.

Dilysss committed:
No assert statements

To include assertions to document important assumptions that
should hold at various points in the code

Add assert statements into markDone, markUndone, addToDo,
addEvent and addDeadline methods

Using these assertions allows us to ensure that the
respective tasks are marked as done or undone. Adding
these assertions also allows us to ensure that the todo,
events and deadlines are added to the tasklist

Dilysss committed:
Improve Code Quality

Dilysss committed:
Merge pull request #1 from Dilysss/branch-A-Assertions

Add Assertions

Dilysss committed:
Merge branch 'master' into branch-A-CodeQuality

Dilysss committed:
Merge pull request #2 from Dilysss/branch-A-CodeQuality

Improve Code Quality

Dilysss committed:
Merging test

Dilysss committed:
Merge pull request #3 from Dilysss/Merging-PRs

Merging test

Dilysss committed:
Amend Statistics Extension

Dilysss committed:
Improve GUI

Dilysss committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	src/main/java/duke/Ui.java

Dilysss committed:
Add User Guide

Dilysss committed:
Amend User Guide

Dilysss committed:
Finalise iP Features

Dilysss committed:
Amend Final Features

Dilysss committed:
Edit User Guide

Dilysss committed:
Amend UG

Dilysss committed:
Edit UG
